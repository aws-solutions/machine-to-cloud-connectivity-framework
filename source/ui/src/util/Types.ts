// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import React from 'react';

export type ErrorMessage = { errorMessage: string };
export type FormControlElement = HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement;
export type KeyStringValue = Record<string, string | undefined>;
type ApiMethod = 'get' | 'post';
type GreengrassCoreDeviceStatus = 'HEALTHY' | 'UNHEALTHY';
type LoggingLevel = 'ERROR' | 'WARN' | 'INFO' | 'DEBUG' | 'VERBOSE';

export enum ConnectionControl {
  DEPLOY = 'deploy',
  START = 'start',
  STOP = 'stop',
  UPDATE = 'update',
  DELETE = 'delete',
  PUSH = 'push',
  PULL = 'pull',
  FAIL = 'fail'
}

export enum LogType {
  INFO = 'info',
  ERROR = 'error'
}

export enum MachineProtocol {
  OPCDA = 'opcda',
  OPCUA = 'opcua',
  OSIPI = 'osipi',
  MODBUSTCP = 'modbustcp'
}

export enum MessageModalType {
  MESSAGE,
  CONFIRM
}

export enum PaginationType {
  PREV,
  NEXT
}

export enum CreatedBy {
  SYSTEM = 'System',
  USER = 'User'
}

export enum OsPlatform {
  LINUX = 'linux',
  WINDOWS = 'windows'
}

export enum GreengrassCoreDeviceControl {
  CREATE = 'create',
  DELETE = 'delete'
}

export enum OsiPiAuthMode {
  ANONYMOUS = 'ANONYMOUS',
  BASIC = 'BASIC'
  //TODO: Figure out how to support Kerberos auth
  // KERBEROS = 'KERBEROS'
}

export interface AmplifyConfigurationInput {
  apiEndpoint: string;
  identityPoolId: string;
  loggingLevel: LoggingLevel;
  region: string;
  s3Bucket: string;
  userPoolId: string;
  webClientId: string;
}

export interface AmplifyConfigurationResponse {
  API: {
    endpoints: { name: string; endpoint: string; region: string }[];
  };
  Auth: {
    identityPoolId: string;
    region: string;
    userPoolId: string;
    userPoolWebClientId: string;
  };
  Storage: {
    AWSS3: {
      bucket: string;
      region: string;
    };
  };
}

export interface ConnectionDefinition {
  control: ConnectionControl;
  connectionName: string;
  protocol: MachineProtocol;
  area?: string;
  greengrassCoreDeviceName?: string;
  machineName?: string;
  opcDa?: OpcDaDefinition;
  opcUa?: OpcUaDefinition;
  osiPi?: OsiPiDefinition;
  modbusTcp?: ModbusTcpDefinition;
  process?: string;
  sendDataToIoTSiteWise?: boolean;
  sendDataToIoTTopic?: boolean;
  sendDataToKinesisDataStreams?: boolean;
  sendDataToTimestream?: boolean;
  sendDataToHistorian?: boolean;
  historianKinesisDatastreamName?: string;
  siteName?: string;
  logLevel?: string;
}

export interface CommonDefinition {
  machineIp: string;
  serverName: string;
}

export interface OpcDaDefinition extends CommonDefinition {
  interval: number | string;
  iterations: number | string;
  listTags?: string[];
  tags?: string[];
}

export interface OpcUaDefinition extends CommonDefinition {
  port?: number | string;
}

export interface OsiPiDefinition {
  apiUrl: string;
  serverName: string;
  authMode: OsiPiAuthMode;
  verifySSL: boolean;
  username?: string;
  password?: string;
  tags: string[];
  requestFrequency: number | string;
  catchupFrequency: number | string;
  maxRequestDuration: number | string;
  queryOffset: number | string;
}

export interface ModbusTcpDefinition {
  host: string;
  hostPort: number | string;
  hostTag: string;
  modbusSlavesConfigSerialized: string;
  modbusSlavesConfig: ModbusTcpSlaveDefinition[];
}

export interface ModbusTcpSlaveDefinition {
  slaveAddress: number | string;
  frequencyInSeconds: number;
  commandConfig: ModbusTcpSlaveCommandConfig;
}

export interface ModbusTcpSlaveCommandConfig {
  readCoils?: ModbusTcpSlaveCommandReadCoilsConfig;
  readDiscreteInputs?: ModbusTcpSlaveCommandReadDiscreteInputsConfig;
  readHoldingRegisters?: ModbusTcpSlaveCommandReadHoldingRegistersConfig;
  readInputRegisters?: ModbusTcpSlaveCommandReadInputRegistersConfig;
}

export interface ModbusTcpSlaveCommandIndividualConfig {
  address: number | string;
  count?: number | string;
}

export type ModbusTcpSlaveCommandReadCoilsConfig = ModbusTcpSlaveCommandIndividualConfig;
export type ModbusTcpSlaveCommandReadDiscreteInputsConfig = ModbusTcpSlaveCommandIndividualConfig;
export type ModbusTcpSlaveCommandReadHoldingRegistersConfig = ModbusTcpSlaveCommandIndividualConfig;
export type ModbusTcpSlaveCommandReadInputRegistersConfig = ModbusTcpSlaveCommandIndividualConfig;

export interface ListConnectionsItem {
  connectionName: string;
  machineName: string;
  protocol: MachineProtocol;
  status: ConnectionControl;
  sendDataToIoTSiteWise: boolean;
  sendDataToIoTTopic: boolean;
  sendDataToKinesisDataStreams: boolean;
  sendDataToHistorian: boolean;
}

export interface ListConnectionsResponse {
  connections: ListConnectionsItem[];
  nextToken?: string;
}

export interface ControlConnectionResponse {
  connectionName: string;
  control: ConnectionControl;
  result: string;
}

export interface CreateUpdateConnectionResponse {
  connectionName: string;
  control: ConnectionControl.DEPLOY | ConnectionControl.UPDATE;
  message: string;
}

export interface ModalProps {
  show: boolean;
  hide: () => void;
}

export interface ConnectionLogsModalProps extends ModalProps {
  connectionName: string;
}

export interface MessageModalProps extends ModalProps {
  message: string | React.ReactNode;
  modalType: MessageModalType;
  confirmAction?: () => void | Promise<void>;
}

export interface GetConnectionResponse {
  connectionName: string;
  greengrassCoreDeviceName: string;
  protocol: MachineProtocol;
  sendDataToIoTSiteWise: boolean;
  sendDataToIoTTopic: boolean;
  sendDataToKinesisDataStreams: boolean;
  sendDataToTimestream: boolean;
  sendDataToHistorian: boolean;
  area?: string;
  machineName?: string;
  logLevel?: string;
  opcDa?: OpcDaDefinition;
  opcUa?: OpcUaDefinition;
  osiPi?: OsiPiDefinition;
  modbusTcp?: ModbusTcpDefinition;
  process?: string;
  siteName?: string;
  opcDaListTags?: string;
  opcDaTags?: string;
  osiPiTags?: string;
  historianKinesisDatastreamName?: string;
}

export interface FormProps {
  connection: GetConnectionResponse;
  onChange: (event: React.ChangeEvent<FormControlElement>) => void;
  errors: KeyStringValue;
}

export interface OpcUaFormProps extends FormProps {
  region: string;
}

export interface ListLogsItem {
  connectionName: string;
  timestamp: number;
  logType: LogType;
  message: string;
}

export interface ListLogsResponse {
  logs: ListLogsItem[];
  nextToken?: string;
}

export interface AlphaNumericValidationProps {
  value: string;
  maxLength: number;
  errors: KeyStringValue;
  errorKeyName: string;
  errorMessage: string;
}

export interface ErrorProps {
  message?: string;
  response?: {
    data: ErrorMessage | unknown;
  };
}

export interface ApiProps {
  method: ApiMethod;
  path: string;
  options?: Record<string, unknown>;
}

export interface ListGreengrassCoreDevicesResponse {
  greengrassCoreDevices: ListGreengrassCoreDevicesItem[];
  nextToken?: string;
}

export interface ListGreengrassCoreDevicesItem {
  name: string;
  createdBy: CreatedBy;
  numberOfConnections: number;
  osPlatform: string;
}

export interface UserGreengrassCoreDeviceResponse {
  greengrassCoreDevices: UserGreengrassCoreDevice[];
}

export interface UserGreengrassCoreDevice {
  coreDeviceThingName: string;
  status: GreengrassCoreDeviceStatus;
}

export interface GreengrassCoreDevicePostResponse {
  name: string;
  message: string;
}

export interface GreengrassCoreDeviceDomIds {
  category: string;
  greengrassCoreDeviceName: string;
  osPlatform: string;
}
