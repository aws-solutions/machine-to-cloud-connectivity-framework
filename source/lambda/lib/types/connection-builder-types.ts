// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  CreatedBy,
  GetConnectionResponse,
  GetConnectionsResponse,
  GetGreengrassCoreDevicesResponse,
  GetLogsResponse
} from './dynamodb-handler-types';
import { ListGreengrassCoreDevicesResponse } from './greengrass-v2-handler-types';
import { ConnectionControl } from './solution-common-types';

type StringJson = Record<string, string>;
type StringArrayJson = Record<string, string[]>;
export type APIResponseBodyType =
  | GetConnectionsResponse
  | GetConnectionResponse
  | GetGreengrassCoreDevicesResponse
  | GetLogsResponse
  | ListGreengrassCoreDevicesResponse
  | ProcessConnectionResponse
  | ProcessGreengrassCoreDeviceResponse
  | Record<string, never>;

export interface APIGatewayRequest {
  resource: string;
  path: string;
  httpMethod: string;
  headers: StringJson;
  multiValueHeaders: StringArrayJson;
  queryStringParameters: StringJson;
  multiValueQueryStringParameters: StringArrayJson;
  pathParameters: StringJson;
  stageVariables: StringJson;
  requestContext: Record<string, object>;
  body: string;
  isBase64Encoded: string;
}

export interface APIGatewayResponse {
  statusCode: number;
  body: string;
  headers?: StringJson;
  multiValueHeaders?: StringArrayJson;
  isBase64Encoded?: boolean;
}

export interface ProcessConnectionResponse {
  connectionName: string;
  control: ConnectionControl;
  message: string;
}

export interface GetApiRequestInput {
  path: string;
  pathParameters: StringJson;
  queryStrings: StringJson;
  resource: string;
}

export interface PostApiRequestInput {
  body: string;
  resource: string;
}

export enum GreengrassCoreDeviceControl {
  CREATE = 'create',
  DELETE = 'delete'
}

export enum GreengrassCoreDeviceEventTypes {
  CREATE = 'CreateGreengrassCoreDevice',
  DELETE = 'DeleteGreengrassCoreDevice'
}

export interface PostGreengrassRequestBodyInput {
  name: string;
  control: GreengrassCoreDeviceControl;
  createdBy: CreatedBy;
}

export interface ProcessGreengrassCoreDeviceResponse {
  name: string;
  message: string;
}
