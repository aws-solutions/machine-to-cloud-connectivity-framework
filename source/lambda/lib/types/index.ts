// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export * as ConnectionBuilderTypes from './connection-builder-types';
export * as CustomResourceTypes from './custom-resource-types';
export * as DynamoDBHandlerTypes from './dynamodb-handler-types';
export * as GreengrassDeployerTypes from './greengrass-deployer-types';
export * as GreengrassV2HandlerTypes from './greengrass-v2-handler-types';
export * as IoTHandlerTypes from './iot-handler-types';
export * as IoTSiteWiseHandlerTypes from './iot-sitewise-handler-types';
export * as MessageConsumerTypes from './message-consumer-types';
export * as S3HandlerTypes from './s3-handler-types';
export * as SolutionCommonTypes from './solution-common-types';
export * as UtilsTypes from './utils-types';
