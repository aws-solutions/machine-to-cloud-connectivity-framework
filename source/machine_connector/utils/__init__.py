# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from .client import AWSEndpointClient
from .pickle_checkpoint_manager import PickleCheckpointManager
from .stream_manager_helper import StreamManagerHelperClient
from .init_msg_metadata import InitMessage

<<<<<<< HEAD
__version__ = "4.1.0"
=======
__version__ = "4.2.2"
>>>>>>> main
