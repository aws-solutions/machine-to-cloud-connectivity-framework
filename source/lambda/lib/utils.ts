// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import axios from 'axios';
import { init } from '@paralleldrive/cuid2';
import { LambdaError } from './errors';
import { AnonymousMetricData, AwsSdkOptions } from './types/utils-types';

const METRICS_ENDPOINT = 'https://metrics.awssolutionsbuilder.com/generic';

/**
 * Generates an unique ID based on the parameter length.
 * @param length The length of the unique ID
 * @returns The unique ID
 */
export function generateUniqueId(length: number = 4) {
  const createId = init({
    length: length
  });
  return createId().toUpperCase();
}

/**
 * Gets the AWS JavaScript SDK options.
 * If the solution ID and version are provided, the options includes the custom user agent.
 * @param input The optional AWS SDK options
 * @returns The custom user agent options for AWS JavaScript SDK
 */
export function getAwsSdkOptions(input?: AwsSdkOptions): AwsSdkOptions {
  const { SOLUTION_ID, SOLUTION_VERSION } = process.env;
  const options: AwsSdkOptions = { ...input };

  if (SOLUTION_ID && SOLUTION_VERSION && SOLUTION_ID.trim() !== '' && SOLUTION_VERSION.trim() !== '') {
    options.customUserAgent = `AwsSolution/${SOLUTION_ID}/${SOLUTION_VERSION}`;
  }

  return options;
}

/**
 * Sends anonymous usage metrics.
 * @param data Data to send a anonymous metric
 * @param uuid The solution UUID
 */
export async function sendAnonymousMetric(data: AnonymousMetricData, uuid: string) {
  try {
    const { SOLUTION_ID, SOLUTION_VERSION } = process.env;
    const body = {
      Solution: SOLUTION_ID,
      Version: SOLUTION_VERSION,
      UUID: uuid,
      TimeStamp: new Date().toISOString().replace('T', ' ').replace('Z', ''),
      Data: data
    };

    const config = {
      headers: { 'Content-Type': 'application/json' }
    };

    await axios.post(METRICS_ENDPOINT, JSON.stringify(body), config);
  } catch (error) {
    console.error('Error sending an anonymous metric: ', error);
  }
}

/**
 * Sleeps for seconds.
 * @param seconds The seconds to sleep
 * @returns Promise to sleep for seconds
 */
export async function sleep(seconds: number): Promise<PromiseConstructor> {
  return new Promise(resolve => {
    setTimeout(resolve, seconds * 1000);
  });
}

/**
 * Trims all strings in the object or array.
 * @param obj The object to trim all strings
 * @returns The trimmed strings in the object
 * @throws `UtilsInvalidObject` when object is not provided
 */
export function trimAllStringInObjectOrArray(obj: unknown) {
  if (typeof obj !== 'object') {
    throw new LambdaError({
      message: 'Invalid object',
      name: 'UtilsInvalidObject',
      statusCode: 400
    });
  }

  let trimmedObject: unknown;
  if (Array.isArray(obj)) {
    trimmedObject = [];
  } else {
    trimmedObject = {};
  }

  for (const key in obj) {
    if (typeof obj[key] === 'object') {
      trimmedObject[key] = trimAllStringInObjectOrArray(obj[key]);
    } else {
      if (typeof obj[key] === 'string') {
        obj[key] = obj[key].trim();
      }

      if (Array.isArray(trimmedObject)) {
        trimmedObject.push(obj[key]);
      } else {
        trimmedObject[key] = obj[key];
      }
    }
  }

  return trimmedObject;
}

/**
 * Validates if the provided version is valid.
 * The valid version consists of a major version number, a minor version number, and a patch version number.
 * (e.g. 1.0.0, v1.0.0)
 * @param version The version to validate
 * @returns If the version is valid or not
 */
export function isValidVersion(version: string): boolean {
  return (
    /^(\d|[1-9]\d*)\.(\d|[1-9]\d*)\.(\d|[1-9]\d*)$/.test(version) ||
    /^v(\d|[1-9]\d*)\.(\d|[1-9]\d*)\.(\d|[1-9]\d*)$/.test(version)
  );
}

/**
 * Extracts the version from the version string.
 * The expected solution version is vX.Y.Z. X, Y, and Z are integer numbers.
 * When the solution version is v1.0.0, the component version would be 1.0.0.
 * If the solution version is not expected format (e.g. customVersion, v.1.0.0, v1, and so on),
 * the Lambda function is going to replace the component version to `1.0.0` by default.
 * @param version
 * @returns The version number
 */
export function extractValidVersion(version: string): string {
  if (isValidVersion(version)) {
    return version.replace(/^v/, '');
  } else {
    return '1.0.0';
  }
}
