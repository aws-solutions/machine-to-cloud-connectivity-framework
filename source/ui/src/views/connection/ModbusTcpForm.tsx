// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { I18n } from '@aws-amplify/core';
import Form from 'react-bootstrap/Form';
import { FormProps } from '../../util/types';

/**
 * Renders the Modbus TCP form.
 * @param props The properties for the Modbus TCP form
 * @returns The Modbus TCP form
 */
export default function ModbusTcpForm(props: FormProps): React.JSX.Element {
  const { connection, onChange, errors } = props;

  return (
    <>
      <Form.Group>
        <Form.Label>
          {I18n.get('host')} <span className="red-text">*</span>
        </Form.Label>
        <Form.Text muted>{I18n.get('description.modbus.tcp.host')}</Form.Text>
        <Form.Control
          id="modbusTcp_host"
          required
          type="text"
          placeholder={I18n.get('placeholder.modbus.tcp.host')}
          defaultValue={connection.modbusTcp?.host}
          onChange={onChange}
          isInvalid={errors.modbusTcp_host != undefined}
        />
        <Form.Control.Feedback type="invalid">{errors.modbusTcp_host}</Form.Control.Feedback>
      </Form.Group>
      <Form.Group>
        <Form.Label>
          {I18n.get('port')} <span className="red-text">*</span>
        </Form.Label>
        <Form.Text muted>{I18n.get('description.modbus.tcp.hostPort')}</Form.Text>
        <Form.Control
          id="modbusTcp_hostPort"
          required
          type="text"
          placeholder={I18n.get('placeholder.modbus.tcp.hostPort')}
          defaultValue={connection.modbusTcp?.hostPort}
          onChange={onChange}
          isInvalid={!!errors.modbusTcp_hostPort}
        />
        <Form.Control.Feedback type="invalid">{errors.modbusTcp_hostPort}</Form.Control.Feedback>
      </Form.Group>
      <Form.Group>
        <Form.Label>
          {I18n.get('host.tag')} <span className="red-text">*</span>
        </Form.Label>
        <Form.Text muted>{I18n.get('description.modbus.tcp.hostTag')}</Form.Text>
        <Form.Control
          id="modbusTcp_hostTag"
          required
          type="text"
          placeholder={I18n.get('placeholder.modbus.tcp.hostTag')}
          defaultValue={connection.modbusTcp?.hostTag}
          onChange={onChange}
          isInvalid={!!errors.modbusTcp_hostTag}
        />
        <Form.Control.Feedback type="invalid">{errors.modbusTcp_hostTag}</Form.Control.Feedback>
      </Form.Group>
      <Form.Group>
        <Form.Label>
          {I18n.get('modbus.secondary.config')} <span className="red-text">*</span>
        </Form.Label>
        <Form.Text muted>{I18n.get('description.modbus.secondary.config')}</Form.Text>
        <Form.Control
          id="modbusTcp_modbusSecondariesConfigSerialized"
          required
          type="text"
          placeholder={I18n.get('placeholder.modbus.secondary.config')}
          defaultValue={connection.modbusTcp?.modbusSecondariesConfigSerialized}
          onChange={onChange}
          isInvalid={!!errors.modbusTcp_modbusSecondariesConfigSerialized}
          as="textarea"
          rows={15}
        />
        <Form.Control.Feedback type="invalid">
          {errors.modbusTcp_modbusSecondariesConfigSerialized}
        </Form.Control.Feedback>
      </Form.Group>
    </>
  );
}
