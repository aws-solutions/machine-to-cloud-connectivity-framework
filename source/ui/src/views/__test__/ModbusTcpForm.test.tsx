// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { I18n } from '@aws-amplify/core';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ModbusTcpForm from '../connection/ModbusTcpForm';
import { GetConnectionResponse, MachineProtocol } from '../../util/types';
import { INIT_CONNECTION } from '../../util/utils';

test('renders the default ModbusTcpForm', async () => {
  const connection: GetConnectionResponse = INIT_CONNECTION;
  const modbusTcpForm = render(
    <ModbusTcpForm connection={connection} onChange={() => console.log('changed')} errors={{}} />
  );
  expect(modbusTcpForm.container).toMatchSnapshot();
});

test('renders the ModbusTcpForm with the existing connection', async () => {
  const connection: GetConnectionResponse = {
    connectionName: 'mock-connection-id',
    greengrassCoreDeviceName: 'mock-greengrass-device-name',
    protocol: MachineProtocol.MODBUSTCP,
    sendDataToIoTSiteWise: true,
    sendDataToIoTTopic: true,
    sendDataToKinesisDataStreams: true,
    sendDataToTimestream: false,
    sendDataToHistorian: true,
    area: 'mock-area',
    machineName: 'mock-machine',
    modbusTcp: {
      host: 'mock-host',
      hostPort: 1,
      hostTag: 'mock-tag',
      modbusSecondariesConfigSerialized: 'mock-serialized',
      modbusSecondariesConfig: [
        {
          secondaryAddress: 1,
          frequencyInSeconds: 1,
          commandConfig: {
            readCoils: {
              address: 1,
              count: 1
            }
          }
        }
      ]
    }
  };

  const modbusTcpForm = render(
    <ModbusTcpForm connection={connection} onChange={() => console.log('changed')} errors={{}} />
  );
  expect(modbusTcpForm.container).toMatchSnapshot();
});

test('renders the ModbusTcpForm errors', async () => {
  const connection: GetConnectionResponse = INIT_CONNECTION;
  const errors = {
    connectionName: 'connectionName invalid',
    sendDataTo: 'sendDataTo invalid',
    modbusTcp_host: 'modbusTcp_host invalid',
    modbusTcp_hostPort: 'modbusTcp_hostPort invalid',
    modbusTcp_hostTag: 'modbusTcp_hostTag invalid',
    modbusTcp_modbusSecondariesConfigSerialized: 'modbusTcp_modbusSecondariesConfigSerialized invalid'
  };
  const modbusTcpForm = render(
    <ModbusTcpForm connection={connection} onChange={() => console.log('changed')} errors={errors} />
  );
  expect(modbusTcpForm.container).toMatchSnapshot();
});

test('tests handleValueChange function', async () => {
  const connection: GetConnectionResponse = INIT_CONNECTION;
  const modbusTcpForm = render(
    <ModbusTcpForm connection={connection} onChange={() => console.log('changed')} errors={{}} />
  );
  const input = modbusTcpForm.getByPlaceholderText(I18n.get('placeholder.modbus.tcp.host'));
  userEvent.type(input, '127.0.0.1');
  expect((input as HTMLInputElement).value).toEqual('127.0.0.1');
});
