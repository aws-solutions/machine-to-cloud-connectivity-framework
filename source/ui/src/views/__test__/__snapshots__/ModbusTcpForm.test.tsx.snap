// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the ModbusTcpForm errors 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      host
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.host
    </small>
    <input
      class="form-control is-invalid"
      id="modbusTcp_host"
      placeholder="placeholder.modbus.tcp.host"
      required=""
      type="text"
      value=""
    />
    <div
      class="invalid-feedback"
    >
      modbusTcp_host invalid
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      port
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.hostPort
    </small>
    <input
      class="form-control is-invalid"
      id="modbusTcp_hostPort"
      placeholder="placeholder.modbus.tcp.hostPort"
      required=""
      type="text"
      value="502"
    />
    <div
      class="invalid-feedback"
    >
      modbusTcp_hostPort invalid
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      host.tag
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.hostTag
    </small>
    <input
      class="form-control is-invalid"
      id="modbusTcp_hostTag"
      placeholder="placeholder.modbus.tcp.hostTag"
      required=""
      type="text"
      value=""
    />
    <div
      class="invalid-feedback"
    >
      modbusTcp_hostTag invalid
    </div>
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      modbus.secondary.config
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.secondary.config
    </small>
    <textarea
      class="form-control is-invalid"
      id="modbusTcp_modbusSecondariesConfigSerialized"
      placeholder="placeholder.modbus.secondary.config"
      required=""
      rows="15"
      type="text"
    >
      [
    {
        "secondaryAddress": 1,
        "frequencyInSeconds": 30,
        "commandConfig": {
            "readCoils": {
                "address": 1,
                "count": 1
            },
            "readDiscreteInputs": {
                "address": 1
            },
            "readHoldingRegisters": {
                "address": 1
            },
            "readInputRegisters": {
                "address": 1
            }
        }
    }
]
    </textarea>
    <div
      class="invalid-feedback"
    >
      modbusTcp_modbusSecondariesConfigSerialized invalid
    </div>
  </div>
</div>
`;

exports[`renders the ModbusTcpForm with the existing connection 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      host
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.host
    </small>
    <input
      class="form-control"
      id="modbusTcp_host"
      placeholder="placeholder.modbus.tcp.host"
      required=""
      type="text"
      value="mock-host"
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      port
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.hostPort
    </small>
    <input
      class="form-control"
      id="modbusTcp_hostPort"
      placeholder="placeholder.modbus.tcp.hostPort"
      required=""
      type="text"
      value="1"
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      host.tag
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.hostTag
    </small>
    <input
      class="form-control"
      id="modbusTcp_hostTag"
      placeholder="placeholder.modbus.tcp.hostTag"
      required=""
      type="text"
      value="mock-tag"
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      modbus.secondary.config
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.secondary.config
    </small>
    <textarea
      class="form-control"
      id="modbusTcp_modbusSecondariesConfigSerialized"
      placeholder="placeholder.modbus.secondary.config"
      required=""
      rows="15"
      type="text"
    >
      mock-serialized
    </textarea>
    <div
      class="invalid-feedback"
    />
  </div>
</div>
`;

exports[`renders the default ModbusTcpForm 1`] = `
<div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      host
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.host
    </small>
    <input
      class="form-control"
      id="modbusTcp_host"
      placeholder="placeholder.modbus.tcp.host"
      required=""
      type="text"
      value=""
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      port
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.hostPort
    </small>
    <input
      class="form-control"
      id="modbusTcp_hostPort"
      placeholder="placeholder.modbus.tcp.hostPort"
      required=""
      type="text"
      value="502"
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      host.tag
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.tcp.hostTag
    </small>
    <input
      class="form-control"
      id="modbusTcp_hostTag"
      placeholder="placeholder.modbus.tcp.hostTag"
      required=""
      type="text"
      value=""
    />
    <div
      class="invalid-feedback"
    />
  </div>
  <div
    class="form-group"
  >
    <label
      class="form-label"
    >
      modbus.secondary.config
       
      <span
        class="red-text"
      >
        *
      </span>
    </label>
    <small
      class="form-text text-muted"
    >
      description.modbus.secondary.config
    </small>
    <textarea
      class="form-control"
      id="modbusTcp_modbusSecondariesConfigSerialized"
      placeholder="placeholder.modbus.secondary.config"
      required=""
      rows="15"
      type="text"
    >
      [
    {
        "secondaryAddress": 1,
        "frequencyInSeconds": 30,
        "commandConfig": {
            "readCoils": {
                "address": 1,
                "count": 1
            },
            "readDiscreteInputs": {
                "address": 1
            },
            "readHoldingRegisters": {
                "address": 1
            },
            "readInputRegisters": {
                "address": 1
            }
        }
    }
]
    </textarea>
    <div
      class="invalid-feedback"
    />
  </div>
</div>
`;
